Add muon-catalyzed fusion process/model skeletons

stognini opened the pull request:
This PR adds the basic structure for the muon-catalyzed fusion process. See issue [#1508](https://github.com/celeritas-project/celeritas/issues/1508) for details.

 It is an at rest process, and it applies only to d-t mixtures. The model `executor` is responsible for taking a muon at rest and:
- Create the muonic atom
- Apply any secondary effect (e.g. isotopic transfer, where the muon transfers to a heavier atom)
- Create the muonic molecule
- Verify if fusion happens before the muon decays
- Call appropriate interactor (decay or fusion)

stognini committed:
Add initial muon CDF data

stognini committed:
Add process/model skeletons

stognini committed:
Draft host/device data

stognini committed:
Add model implementation; draft host/device data initialization

stognini committed:
Add mucf executor + interactors + helper classes skeletons

stognini committed:
Make sure it builds

stognini committed:
Expand documentation

stognini committed:
Fixup

sethrj commented on src/celeritas/inp/MucfPhysicsData.hh:
```diff
@@ -0,0 +1,117 @@
+ *
+ * \todo This data may be loaded from an external file in the future.
+ */
+static Grid mucf_muon_energy_cdf()
```
These should probably be inlined directly into `MucfPhysics::from_default`. At any rate, they should live inside that cc file.

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+               && muonic_helium3;
+    }
+
+    //! Assign from ParticleParams (anonymous free function in the model.cc?)
```
Correct, this should be in the model.cc file.

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+    if (!pids.MEMBER)                                \
+    {                                                \
+        missing.push_back({#MEMBER, pdg::MEMBER()}); \
+    }
```
Since this doesn't require any code-to-text macro preprocessing (which similar JSON code does, `"X" = X`) I think it would improve readability to make this a lambda function.

```suggestion
```

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+        MP_ADD(mu_minus);
+        MP_ADD(neutron);
+        MP_ADD(proton);
+        MP_ADD(alpha);
```
```suggestion


   auto set_pid = [&](ParticleId& pid, PdgNumber num) {
...
    };
        set_pid(pids.mu_minus, pdg::pids.mu_minus());
...
```

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+            << join(missing.begin(), missing.end(), ", ", [](PairStrPdg const& p) {
+                   return p.first + " (PDG "
+                          + std::to_string(p.second.unchecked_get()) + ')';
+               }));
```
```suggestion
            << join_stream(missing.begin(), missing.end(), ", ", [](std::ostream& os, PairStrPdg const& p) {
                   os << p.first << " (PDG " << p.second << ')';
               }));
```
(include `corecel/OpaqueIdIO.hh`)

stognini replied:
Wait, but without the macro I lose the ability to store the particle name as a string in the lambda function (we don't have a `to_string` to convert our PDGs to their names to populate the `first` element of the `PairStrPdg`). Am I missing something? 

We can always just print out the PDG, but aside from the basic elementary particles, the PDG numbering scheme gets very cumbersome very fast (that is why heavier particles, like a deuteron, have wildly large PDG numbers).

sethrj replied:
Oh you're right, I did miss that. The macro makes more sense now :)

sethrj replied:
PS the two previous discussions are resolved; but you should still use `join_stream` to simplify the `join` operation.

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+    using MatCycleItems = Collection<T, W, M, MuCfMatCompId>;
+    template<class T>
+    using MaterialItems = Collection<T, W, M, MuCfMatCompId>;
+    using Grid = NonuniformGridRecord;
```
Since the grid and the grid record are different concepts maybe:
```suggestion
    using GridRecord = NonuniformGridRecord;
```

Same thing with ParticleIds vs Particles. I think it's still OK to keep the names as "grid" and "particle" but the type name should reflect the type.

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,156 @@
+    template<class T>
+    using MatCycleItems = Collection<T, W, M, MuCfMatCompId>;
+    template<class T>
+    using MaterialItems = Collection<T, W, M, MuCfMatCompId>;
```
These are the same type, but it looks like one is mat*molecule*spin whereas one is mat? They should have different IDs if the IDs have different semantics.

stognini replied:
Good catch. I did not clean this up while deciding how to organize the data. I'll keep `MaterialItems` for both, since the main index is still the material component: `matcompid_to_matid` allows us to know which component is tied to which `MaterialId`, so from the material id we retrieve all cached material information (for now only cycle time, but this will be equivalent for the other time data, like spin flip).

sethrj commented on src/celeritas/mucf/interactor/DDMucfInteractor.hh:
```diff
@@ -0,0 +1,152 @@
+ * Construct with shared and state data.
+ */
+DDMucfInteractor::DDMucfInteractor(NativeCRef<DTMixMucfData> const& data,
+                                   Channel const channel,
```
```suggestion
                                   Channel channel,
```

sethrj commented on src/celeritas/mucf/interactor/DDMucfInteractor.hh:
```diff
@@ -0,0 +1,152 @@
+/*!
+ * Return number of secondaries from each fusion channel.
+ */
+CELER_FUNCTION size_type DDMucfInteractor::num_secondaries() const
```
This should be a static enum array that lives in the mucf data.

stognini replied:
I pondered about it when drafting the data, and since each molecule (dd, dt, tt) has its own set of channels, keeping the channel list as part of the interactor seemed easier to maintain and follow.

So in your suggestion, do you think we should have hardcoded in the `DTMixMucfData.hh` all channel enums (such as an `enum DDFusionChannel` and so on), and one `EnumArray` for each (e.g. `EnumArray<DDFusionChannel, size_type> dd_num_secondaries;`) in the `struct DTMixMucfData`?

sethrj replied:
I forgot this was  just for DD and that `Channel` was local to the interactor. I think it's ok then, but you shoudl still replace with an EnumArray rather than a `case`.

sethrj commented on src/celeritas/Types.hh:
```diff
@@ -237,7 +243,10 @@ enum class MucfMuonicAtom
+/*!
+ * Muonic molecule selection from material data. This is *not* intended to be
+ * used by the transport loop.
+ */
```
I thought we decided to move all these mucf definitions to a different file (e.g. celeritas/mucf/Types.hh) ? They're very specialized.

stognini replied:
Agreed. Moved everything to a new `celeritas/mucf/Types.hh`.

sethrj commented on src/celeritas/phys/PDGNumber.hh:
```diff
@@ -146,6 +146,10 @@ CELER_DEFINE_PDGNUMBER(anti_he3, -1000020030)
 CELER_DEFINE_PDGNUMBER(alpha, 1000020040)
 CELER_DEFINE_PDGNUMBER(anti_alpha, -1000020040)
 
+// Muonic nuclei
```
Are these numbers standardized or in a "reserved" space? Are they specific to Celeritas or used in Geant4 as well?

stognini replied:
These are standard PDG codes. In G4, they are calculated on the fly for specific cases. E.g. when a muonic atom is created (whenever there is sticking, like a muonic alpha after the fusion happens), the secondary's `G4ParticleDefinition` comes from `G4IonTable::GetMuonicAtom(int Z, int A)`.

sethrj commented on src/celeritas/mucf/model/DTMixMucfModel.hh:
```diff
@@ -0,0 +1,75 @@
+                   ParticleParams const& particles,
+                   MaterialParams const& materials);
+
+    // Particle types and energy ranges that this model applies to
+    SetApplicability applicability() const final;
+
+    // Get the microscopic cross sections for the given particle and material
+    XsTable micro_xs(Applicability) const final;
```
Does this PR depend on #1991? We don't have a proper "at rest" right now.

stognini replied:
It does. We need to select which at rest process happens (decay or mucf) during the stepping loop.
Having said that, since a lot happens in the mucf executor, _if_ mucf is selected, we may still return a decay.

See comments in the executor:
```cpp
52:    // Muon decay may compete against other "actions" in this executor

95:   // Check if muon decays before fusion happens
```

sethrj commented on src/celeritas/mucf/process/MucfProcess.cc:
```diff
@@ -0,0 +1,66 @@
+    //! \todo Fix ImportProcessClass
+    CELER_EXPECT(particles_);
+    CELER_EXPECT(materials_);
+    CELER_EXPECT(particles_->find(pdg::mu_minus()));
```
This will change a runtime error into a debug error (it will be hit before the later validate on the particle type) so I would skip this.

sethrj reviewed (changes requested):
I appreciate the detail classes for the context of long term planning, but I think they should be deleted since they're not used and not tested: having them there is an easy way to introduce zombie code into the codebase.

stognini committed:
Address most comments:
- Move types to new mucf/Types.hh
- Move MucfParticles::from_params to model.cc as a free function
- Move all static data to MucfPhysics.cc
- Refactor/Remove types in DTMixMucfData
- Remove CELER_EXPECT(mu_minus) during MucfProcess construction

stognini commented:
> I appreciate the detail classes for the context of long term planning, but I think they should be deleted since they're not used and not tested

Agreed that they are not tested yet, but they are actively being called in the executor. Do you prefer me to make a todo in the executor and remove them until a future PR?

sethrj commented:
> Agreed that they are not tested yet, but they are actively being called in the executor. Do you prefer me to make a todo in the executor and remove them until a future PR?

Oh I didn't catch that. I'll take another look at your comments later this afternoon.

stognini committed:
Move component id getter function into DTMixMucfData

stognini committed:
Assign all current PDG numbers

sethrj commented on src/celeritas/inp/MucfPhysics.cc:
```diff
@@ -6,12 +6,107 @@
+{
+    //! \todo Implement
+    return {};
+};
```
Unwanted semicolon; also please put all these in an anonymous namespace rather than indivdual `static` (which is the C way of doing the equivalent)

sethrj commented on src/celeritas/inp/MucfPhysics.cc:
```diff
@@ -10,6 +10,103 @@ namespace celeritas
+             0.8814332247557004,
+             0.8970684039087949,
+             0.903583061889251,
+             1.0};
```
Add trailng comma for these long indented lists: 
```suggestion
             1.0,
             };
```

sethrj commented on src/celeritas/inp/MucfPhysics.cc:
```diff
@@ -10,6 +10,103 @@ namespace celeritas
+/*!
+ * Static spin-flip data for muon-catalyzed fusion.
+ *
+ * \todo This data may be loaded from an external file in the future.
```
You can leave off these todos; the functions are all just used for defaulting and can be overridden in the future.

sethrj commented on src/celeritas/inp/MucfPhysics.hh:
```diff
@@ -68,51 +68,60 @@ struct MucfCycleRate
- * thus can be cached at model construction.
+ * state (which is just a \c real_type ) for each combination being needed
+ * in the stepping loop. This is because these rates are material
+ * dependent, and thus can be cached at model construction.
```
vscode is doing some unnecessary reindentation I think :(

sethrj commented on src/celeritas/inp/MucfPhysics.hh:
```diff
@@ -68,51 +68,60 @@ struct MucfCycleRate
- * Muonic atom transfer and muonic atom spin flip are secondary effects and not
- * required for muCF to function.
+ * Muonic atom transfer and muonic atom spin flip are secondary effects and
+ * not required for muCF to function.
```
Here too :(

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,133 @@
+    //! \todo Add mean atom transfer times
+    //!@}
+
+    //! \todo Check whether the data are assigned
```
```suggestion
    //! Check whether the data are assigned
```

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,133 @@
+    //! \todo Check whether the data are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        //! \todo Finalize implementation
```
?

sethrj commented on src/celeritas/mucf/data/DTMixMucfData.hh:
```diff
@@ -0,0 +1,133 @@
+        }
+        // Component ID not found
+        return MuCfMatCompId{};
+    }
```
This should be in a view, it should be labeled `find_` , and note that since we only expect a very few materials to be mucf mat compounds we can do a linear search.

stognini replied:
Yep, I was fine with the loop since we will likely have only 1 material. Maybe 2.

I also considered a view, but seemed an overkill for a single function. It originally was a small lambda inside the executor. Do you think the lambda is okay, or should I go ahead and make the view?

sethrj commented on src/celeritas/mucf/model/DTMixMucfModel.cc:
```diff
@@ -0,0 +1,195 @@
+    make_builder(&host_data.matcompid_to_matid)
+        .insert_back(vec_physmatid.begin(), vec_physmatid.end());
+    make_builder(&host_data.cycle_times)
+        .insert_back(vec_cycle_times.begin(), vec_cycle_times.end());
```
You can skip the intermediate vectors and build directly (also try to use the same names as the data):
```suggestion
    CollectionBuilder cycle_times{&host_data.cycle_times};
    CollectionBuilder matcompid_to_matid{&host_data.matcompid_to_matid};
    for (auto const& matid : range(materials.num_materials()))
    {
        auto const& mat_view = materials.get(PhysMatId{matid});

        // Construct calculator for this material
        detail::DTMixMaterialCalculator mat_calculator(mat_view);
        if (!mat_calculator)
        {
            // Skip non-dt mixture materials
            continue;
        }

        // Store material ID and calculated data
        //! \todo Store mean atom spin flip and transfer times
        matcompid_to_matid.push_back(PhysMatId{matid});
        cycle_times.push_back(mat_calculator.cycle_times());
    }
```

Also you're not really doing a "material component" here, you're doing a "FusionMatId" (or equivalent): it's a separate numbering that's based on the physics materials and what supports fusion.

stognini replied:
Agreed that I can improve the name. Maybe on the lines of mucf material record id? Since it's just the contiguous internal storing of material cached data?

sethrj replied:
I think `MucfMatId` (and you can keep it in the mucf type file) is exactly what you're describing: for optical and physics materials, the mat ID corresponds exactly to a material record.

sethrj commented on src/celeritas/mucf/model/DTMixMucfModel.cc:
```diff
@@ -0,0 +1,195 @@
+
+        // Construct calculator for this material
+        detail::DTMixMaterialCalculator mat_calculator(mat_view);
+        if (!mat_calculator)
```
The operator bool isn't semantically very consistent since this is behaving like a helper rather than a function.

sethrj commented on src/celeritas/mucf/model/detail/DTMixMaterialCalculator.hh:
```diff
@@ -0,0 +1,142 @@
+    static constexpr units::AmuMass tritium()
+    {
+        return units::AmuMass{3.016049281320};
+    }
```
For now, specify these in the `inp` data and provide defaults. Don't hardcode them.

sethrj commented on src/celeritas/mucf/model/detail/DTMixMaterialCalculator.hh:
```diff
@@ -0,0 +1,142 @@
+    static constexpr auto liquid_hydrogen_density()
+    {
+        return units::InvCcDensity{4.25e22};
+    }
```
This should also be input data.

sethrj commented on src/celeritas/mucf/model/detail/DTMixMaterialCalculator.hh:
```diff
@@ -0,0 +1,142 @@
+   }
+ * \endcode
+ */
+class DTMixMaterialCalculator
```
I don't think this class structure makes sense. Maybe you could rebuild as an inserter (e.g., `optical/gen/detail/MatScintSpecInserter.hh`, `em/model/detail/SBTableInserter.hh`, ...) and each call to operator() returns a `FusionMatId`?

stognini replied:
Let me rethink it and see if I can make it more like the inserter.

For me it made sense, because it would calculate and cache multiple quantities per material (cycle times, spin flip times, atom transfer times). And since the vast majority of materials are not a d-t mixture, it seemed reasonable to hide that verification in the calculator: If I create the calculator with a given `MaterialView`, the calculator would tell me via its operator bool if that is a valid fusion material---and if so, that meant I could add all those cached quantities to the model data for the executor.

sethrj replied:
But all these calculations are happening "locally", within a single invocation. There's no need to make a class, precalculate a bunch of stuff, and then make another piece of code insert those through accessors. With an inserter class, you'd do everything locally within `MucfMatId operator()(MatId)`, likely with some private helper functions to calculate these "cached" quantities.

sethrj reviewed (commented):
I would prefer to have at least some level of testing (even if it's running through a bunch of null-ops) here. I see now that the executor is being used, but it would be very helpful to have a code coverage tool tell us what is actually being called and tested.

Maybe it's possible to set up a test that creates a state vector and the model (manually setting all the particle attributes that are needed), and calls the `step` operator for the mucf model?

stognini committed:
Address comments:
- Fix import data
- Refactor the material calculator to behave as a material inserter
- Interactors: define number of secondaries as EnumArrays
- Rename internal material id to MucfMatId
- Remove find MucfMatId function from DTMixMucfData and make it a lambda
  in the executor

stognini committed:
Minor improvements

stognini committed:
Update documentation

stognini committed:
Fixup

stognini committed:
Improve documentation

stognini committed:
Fix `RootInterfaceLinkDef`

sethrj reviewed (approved):
Good enough. I do think we should avoid "skeleton" PRs. I know this is a fairly complex piece of machinery, but in the future let's build up one piece at a time: no skeletons in the code. It's hard to tell if a piece of construction code is correct or optimal if it's not used.

sethrj commented on src/celeritas/mucf/model/DTMixMucfModel.hh:
```diff
@@ -0,0 +1,75 @@
+#include "celeritas/mat/MaterialParams.hh"
+#include "celeritas/mucf/data/DTMixMucfData.hh"
+#include "celeritas/mucf/executor/DTMixMucfExecutor.hh"  // IWYU pragma: associated
+#include "celeritas/phys/InteractionApplier.hh"  // IWYU pragma: associated
```
Not sure these are needed; executor should be in .cc/cu for now

sethrj commented on src/celeritas/mucf/process/MucfProcess.hh:
```diff
@@ -0,0 +1,60 @@
+
+#include "celeritas/Types.hh"
+#include "celeritas/mat/MaterialParams.hh"
+#include "celeritas/phys/ParticleParams.hh"
```
Forward declare these; I think they're breaking the CUDA builds.

stognini committed:
Merge branch 'develop' into mucf-interactor

amandalund commented on src/celeritas/mucf/process/MucfProcess.hh:
```diff
@@ -0,0 +1,60 @@
+    EnergyLossGrid energy_loss(Applicability) const final;
+
+    //! Whether the integral method can be used to sample interaction length
+    bool supports_integral_xs() const final { return true; }  //! \todo Check
```
If this process/model _only_ applies at rest I think this should be false.

amandalund commented on src/celeritas/mucf/process/MucfProcess.hh:
```diff
@@ -0,0 +1,60 @@
+    EnergyLossGrid energy_loss(Applicability) const final;
+
+    //! Whether the integral method can be used to sample interaction length
+    bool supports_integral_xs() const final { return true; }  //! \todo Check
```
If this process/model only applies at rest I think this should be false.

stognini replied:
Good catch. Thanks.

stognini committed:
Address comments:
- Fix integral_xs bool
- Fwd declare material and particle params
- Remove trailing includes

sethrj commented:
@stognini looks like you're missing some `CELER_FUNCTION` decorators:
```
  /__w/celeritas/celeritas/src/celeritas/mucf/executor/DTMixMucfExecutor.hh(56): error #20014-D: calling a __host__ function from a __host__ __device__ function is not allowed
        detail::DTMixMuonicAtomSelector form_atom;
                                        ^
  
  Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"
  
  /__w/celeritas/celeritas/src/celeritas/mucf/executor/DTMixMucfExecutor.hh(69): error #20014-D: calling a __host__ function from a __host__ __device__ function is not allowed
        detail::DTMixMuonicMoleculeSelector form_muonic_molecule;
                                            ^
  
  /__w/celeritas/celeritas/src/celeritas/mucf/executor/DTMixMucfExecutor.hh(118): error #20014-D: calling a __host__ function from a __host__ __device__ function is not allowed
                DDMucfInteractor interact(
                                 ^
  
  /__w/celeritas/celeritas/src/celeritas/mucf/executor/DTMixMucfExecutor.hh(125): error #20014-D: calling a __host__ function from a __host__ __device__ function is not allowed
                DTMucfInteractor interact(
                                 ^
  
  /__w/celeritas/celeritas/src/celeritas/mucf/executor/DTMixMucfExecutor.hh(132): error #20014-D: calling a __host__ function from a __host__ __device__ function is not allowed
                TTMucfInteractor interact(
                                 ^
```

stognini committed:
Fix missing CELER_FUNCTION

stognini committed:
Fix device data

sethrj committed:
Fix tidy warnings

sethrj committed:
Use double-precision for input to fix root errors, and un-inline from_default

sethrj committed:
Add operator bool

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into mucf-interactor

sethrj committed:
fixup! Use double-precision for input to fix root errors, and un-inline from_default

sethrj committed:
fixup! Fix tidy warnings

sethrj committed:
fixup! Fix tidy warnings

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into mucf-interactor

sethrj commented:
@amandalund @stognini are we ready to merge this? According to codecov, literally *none* of the new code is tested or even executed; it simply compiles. I really do not like that. It took three weeks to review and fix the *builds*, and we can't even know if it works yet.

Next time we have to put in a big feature, maybe we focus on hardcoding the main use case, then refactoring and refining as we support more fine-grained features. (E.g., start with supporting only DD fusion, one spin, etc., and at least test that it works.) I think right now our muon code has zero coverage at all.

stognini commented:
Agree that the lack of testing is a problem. This also is one single executor. E.g. a discrete EM interaction we were in general able to make the interactor and model in one or two PRs.

One option, if you'd like, is to revert the ordering I was planning. Keep this as draft, and:
- The next PR can be one of the interactors (which can use mock data for a test). And will use the bare minimum from this branch (since the big picture here is mostly set);
- Then spin selection;
- Then material calculation;
- Then we add the process into the stepping loop;

sethrj commented:
The interactor we had the ability to test individually, which was great. I don't want to spend any more time on muons than necessary, but I'd say the next follow-on should be a PR with an integration-level test of muon physics: either in `accel/` or the lower-level `stepper` test.

sethrj commented:
I'm going to merge this out of expedience to reduce the time spent mitigating code rot; @stognini 's next task is to add a unit test that builds the µCF model and compares low level interactor distributions against the G4 code. Then we'll add an integration test that uses muons.
